Support nomad CLI output with JSON and template format

nak3 opened the pull request.

nak3 commented:
For example...

e.g. JSON output (same with HTTP API access)

~~~
$ nomad eval-status -format json a63bbfde
{
    "ID": "a63bbfde-088c-4256-2a8b-bed149e20134",
    "Priority": 50,
    "Type": "service",
    "TriggeredBy": "job-register",
    "JobID": "example",
    "JobModifyIndex": 6,
    "NodeID": "",
    "NodeModifyIndex": 0,
    "Status": "complete",
    "StatusDescription": "",
    "Wait": 0,
    "NextEval": "",
    "PreviousEval": "",
    "BlockedEval": "",
    "FailedTGAllocs": null,
    "CreateIndex": 7,
    "ModifyIndex": 9
}
~~~

e.g. template output

~~~
$ nomad node-status -format template -t "{{.Attributes.arch}}"  8267c455
amd64
~~~

dadgar commented:
@nak3 can you look into the panics? If you give an invalid template it panics.

Also I wonder if it is better to have a `-json` option and `-t` which automatically sets the format to template. I don't think we gain much with the format flag because when would you use format template and not provide the `-t`?

nak3 commented:
Thank you. I updated. If they are alright, I will squash the commits.

dadgar commented:
Yeah that looks good. Okay so some comments. It would be nice if this could work on arrays of elements (aka `nomad node-status -t '{{range .}} {{.Attributes}} {{end}}`. It should be possible!

These all make sense to me except `nomad status` since that command outputs quite a bit (job_summary, evaluations, allocations, etc). So maybe we remove the `-json` and `-t` flag from there. Would like your feedback.

Once we settle on what we are doing, please add docs to the website. And tests

nak3 commented:
Yes, it is possible. You can get the info as below. (I'm sorry, `node-status` had my mistake and I updated.)

~~~
$ nomad node-status  -t '{{range .Attributes}} {{.}}{{end}}' 57b58140
 amd64 1425 Intel(R) Core(TM) i7-3520M CPU @ 2.90GHz 4 5700 1 1.9.1 1 1 OpenJDK Runtime Environment (build 1.8.0_91-b14) openjdk version "1.8.0_91 OpenJDK 64-Bit Server VM (build 25.91-b14, mixed mode) 1 2.4.1 1 1 0.8.5 1.11.0 linux 4.4.8-300.fc23.x86_64 16512901120 'eac5d65fcc2fa5470af2d73f01c143dd28d1123c+CHANGES' 0.4.1dev fedora 23 /sys/fs/cgroup nak3.example.com 127.0.0.1 8166834176 8256450560 tmpfs
 ~~~

Agree about `node status` doesn't need the options. I removed them from `node status`. The options was implemented in `alloc-status` `eval-status` and `node-status` now.

I added the docs. But do we need tests for each commands? I already added `command/data_format_test.go`, though.

dadgar commented:
@nak3 I meant it would be nice if this worked without specify the individual node, alloc, or evaluation. If it could work on all of them! We should probably add this to `nomad inspect` as well. 

I would add a test that if you pass the -json that -t can't be specified as well.

dadgar commented:
For first cut we can leave as is though without support for the lists of elements

dadgar commented on command/alloc_status.go:
```diff
@@ -38,11 +38,17 @@ Alloc Status Options:
     Show full information.
+
+  -json
+    Display information with json format.
```
`Output the allocation in its JSON format.`

Similar for all the others

dadgar commented on command/alloc_status.go:
```diff
@@ -38,11 +38,17 @@ Alloc Status Options:
+    Display information with json format.
+
+  -t
+    Set golang templates format and display information with it.
```
`Format and display allocation using a Go template`

dadgar commented on command/alloc_status.go:
```diff
@@ -130,6 +139,29 @@ func (c *AllocStatusCommand) Run(args []string) int {
+
+		out, err := f.TransformData(alloc)
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("Error transform the data: %s", err))
```
`Error formatting the data`

nak3 commented:
Thank you. (Sorry, I have already included the lists of elements.)

- [x] Support `-json` and `-t` optinos for list elements of `alloc-status`, `eval-status`, `node-status` and `inspect.` (Please see example below)
- [x] Support `-json` and `-t` options for indivisual `alloc-status`, `eval-status`, `node-status`
- [x] Add test to check -json and -t are not specified
- [x] Update documentation and help output

(e.g)
~~~
$ nomad alloc-status -t "{{range .}} {{.ID}} {{end}}"
 b0f97fea-7ec0-a03d-3dbb-f9a771a5d829  518a1a10-7f2f-fc8a-e6e4-1b070749b491  ffa14d59-9faf-a972-b218-5e0005ccd626  95203acf-2409-5c33-eb26-7eb7cd12f37d 
~~~

dadgar commented:
This is pretty awesome. Not sure if this is the best way to do it by this lets you do basic bash autocomplete:

```
$ echo complete.sh
allocs=`nomad alloc-status -t '{{range .}}{{.ID | printf "%s\n"}}{{end}}'`
COMREPLY=( $(compgen -W "${allocs[*]}" -- "${ENTRY}") )
echo $COMREPLY

nomad alloc-status ""
Prefix matched multiple allocations

ID        Eval ID   Job ID   Task Group  Desired Status  Client Status
3b5d8bb3  81ba9bd1  example  cache       run             running
3ecacf5e  81ba9bd1  example  cache       run             running
57d182d0  81ba9bd1  example  cache       run             running
700f01ac  81ba9bd1  example  cache       run             running
803b2685  81ba9bd1  example  cache       run             running
a32028f3  81ba9bd1  example  cache       run             running
bf001125  81ba9bd1  example  cache       run             running
c16540e1  81ba9bd1  example  cache       run             running
f7870dee  81ba9bd1  example  cache       run             running
fe4505da  81ba9bd1  example  cache       run             running

$ ENTRY=c1 ./complete.sh
c16540e1-1d31-ba13-6324-871c1ae35bda
```

We could definitely use this to build an autocomplete

dadgar commented on command/data_format_test.go:
```diff
@@ -0,0 +1,41 @@
+			t.Fatalf("expected output: %s, actual: %s", expectOutput[k], result)
+		}
+	}
+}
```
Can we add a test that you get an error with an invalid go template format

dadgar commented on command/eval_status.go:
```diff
@@ -93,6 +142,35 @@ func (c *EvalStatusCommand) Run(args []string) int {
 		return 1
 	}
+
+	if len(args) == 0 {
```
What is this stuff for?

dadgar commented on command/inspect.go:
```diff
@@ -52,6 +50,50 @@ func (c *InspectCommand) Run(args []string) int {
 		return 1
 	}
 
+	// If args not specified but output format is specified, format and output the jobs data list
```
I think you can clean this one up by removing the -json option since that is true by default.

Further, since the formatter takes `interface{}`, I think we can cleanup the code by doing something like:

```
var toFormat interface{}
if len(args) == 1 {
  toFormat := getJob()
} else {
  toFormat = listJobs()
}

formatter.Transform(toFormat)
```

dadgar commented on command/inspect.go:
```diff
@@ -52,6 +50,50 @@ func (c *InspectCommand) Run(args []string) int {
 
+	// If args not specified but output format is specified, format and output the jobs data list
+	if len(args) == 0 {
+		var format string
```
It should be possible to use the `-template` flag when specifying a particular job

dadgar commented on command/inspect.go:
```diff
@@ -30,20 +30,18 @@ func (c *InspectCommand) Synopsis() string {
 }
 
 func (c *InspectCommand) Run(args []string) int {
+	var ojson bool
```
Docs

nak3 commented:
Thank you @dadgar I updated.

dadgar commented:
Awesome work!
